### ISSUE TITLE
NVDA freezes when using wasapi

### ISSUE THREAD

**Issue by beqabeqa473** [nvaccess/nvda#15752]
cc @jcsteh 
### Steps to reproduce:
There is a way of reproducing this when using Newfon or RHVoice synth drivers
install one of the listed addons.
enter the following code in NVDA console
import synthDriverHandler; [synthDriverHandler.setSynth(synthDriverHandler.getSynth().name) for i in range(20)]
### Actual behavior:
NVDA is freezing and logging lots of errors.
Player is becoming None and idle cannot be called
### Expected behavior:
NVDA should be stable.
#### NVDA installed/portable/running from source:
Running from source, but reproduceable on 2023.3 Release version
#### NVDA version:
Master
#### Windows version:
Windows 10 22h2
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Yes, 2023.3 behaves the same
#### If NVDA add-ons are disabled, is your problem still occurring?
Yes, but easily reproduceable with Newfon or RHVoice synth
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Yes

**Comment by jcsteh** [comment]
A log would be helpful.

When are you seeing this in real usage? I accept that this shouldn't happen, but the code you've provided (while perhaps helpful to reproduce) is pretty extreme and that situation would never actually occur in real usage.

**Comment by beqabeqa473** [comment]
It is happening in real usage, but is not always reproduceable. This is a way to 100% reproduce an issue

**Comment by beqabeqa473** [comment]
It is randomly reproduceable, for example i was talking and didn't touching a keyboard, some large text came to nvda and this bug was reproduced. I cannot give a specific steps when it can be 100% reproduced. But This way is causing the same set of exceptions.

**Comment by beqabeqa473** [comment]
[nvda.log](https://github.com/nvaccess/nvda/files/13278654/nvda.log)

**Comment by a11cf0** [comment]
> It is randomly reproduceable, for example i was talking and didn't touching a keyboard, some large text came to nvda and this bug was reproduced. I cannot give a specific steps when it can be 100% reproduced. But This way is causing the same set of exceptions.

I can confirm this. On my system this happens a lot when working with long passages of text, especially when navigating or quickly interrupting speech. The errors are the same. All synths including eSpeak NG are affected. However, I don't now any other ways to predictably reproduce this, except that described  above.

**Comment by jcsteh** [comment]
> Player is becoming None and idle cannot be called

I don't see this anywhere in your log. Also, this sounds more like #15681, which is now fixed. Do you have another log showing this?

**Comment by beqabeqa473** [comment]
It was only once, when somehow i was able to keep synthesizer talking while logging errors, and with these errors, i discovered a place where None type object has no attribute idle, but now it is not reproduceable anymore. i tried lots of times but no result. It maybe was from another log snippet, but it still was in latest master.
You were saying in your last pr that "Verified that everything still works with this change. I can't reproduce this reliably enough to actually test the workaround in practice." This issue is also very hard to reproduce when you want to do it, but it is reproduceable very often when getting large texts, just navigating through interface. I just found a way, where you could reproduce this issue yourself.

**Comment by jcsteh** [comment]
If you find some way of reproducing this in core, even with Python console code, please let me know. I'd really rather not dabble with add-ons here, partly because it adds another entire piece to an already very complicated puzzle i have to try to understand in order to diagnose the problem.

**Comment by beqabeqa473** [comment]
ok, i was able to reproduce it without addons, but it is still not real usage example. I don't know what to do to ensure this is reproduceable everytime i want.

import speech
from time import sleep


def test(amount):
    for i in range(amount):
        speech.speak(["dsfsfsfdbvcbcvxbcvdfgfdgdfgertretereethghfghbcvb"])
        sleep(0.5)
        speech.cancelSpeech()

test(50)

It is happening with espeak also.

I have no idea for now how to reproduce it with another way, but it is really boring seeing this.

I was not opening an issue, because there were some of them already, and i was hoping this would be fixed with these bugs, but no.

**Comment by jcsteh** [comment]
I wasn't able to reproduce with that snippet unfortunately. Thank you for your efforts, though.

I guess you can disable WASAPI for now, since the toggle is still present. If we continue to see more regressions, it might be worth considering flipping this off by default again in the next release or maybe even reverting it altogether.

**Comment by zstanecic** [comment]
Hi James and Beqa,

I cannot reproduce this too.

**Comment by jcsteh** [comment]
It's not ignoring the problem at all. If WASAPI is causing too many critical issues, no one other than me can fix them, and I don't have sufficient time to diagnose and fix them, it's not really a sustainable situation.

**Comment by beqabeqa473** [comment]
Adjusting volume of sounds is the only thing because of that I am using wasapi. I would be happy to use winmm, if there would be such possibility in winmm player.

Lots of users were waiting to implement of #1409 and reverting of that or continuing to experience this issue is not good at all.

**Comment by cyrmax** [comment]
Another successfull reproduction of this issue. Now with Espeak-NG. Looks like it could be reproduced when typing text or reading text by characters at very fast speed.
Here is another NVDA log, now with all debug messages.
I was just trying to type a command in WSL terminal and NVDA freezed for aprox 5 seconds with the same error.
[nvda2.log](https://github.com/nvaccess/nvda/files/13279846/nvda2.log)

**Comment by zstanecic** [comment]
Well, also when before wasapi, I experienced lags when using Bluetooth headsets. Win mm is not acceptable.

**Comment by jcsteh** [comment]
Well, for now, I cannot fathom how the code can end up in this state. So, until I or someone else can figure it out, we either need to live with this issue or turn it off/revert it so users don't experience the problem. There is literally no other choice.

From what I can see:

1. We put the player into the instances dict as soon as it is created. There can't be an exception between those two points.
2. In the destructor, we remove the player from the instances dict as soon as we destroy it.
3. As soon as we do that, we set the player handle to None. We also return early in the destructor if the handle is None. That means this is not caused by the destructor running twice.

**Comment by zstanecic** [comment]
It can be turned off from the settings. Wasapi is still a part of advanced settings, so I don’t see the problem.

**Comment by jcsteh** [comment]
It looks like a weakref callback might run before `__del__` in some cases; e.g. if the object is garbage collected rather than being destroyed immediately due to its ref count dropping to 0. I don't know why we would garbage collect a WasapiWavePlayer; there shouldn't be any cycles. Still, if I'm correct about the cause of the issue, it's not unreasonable to just ignore the exception.

**Comment by beqabeqa473** [comment]
Then it shouldn't be garbage collected.

**Comment by jcsteh** [comment]
Actually, the correct fix is to just not explicitly remove the key, since it will be implicitly removed by WeakValueDictionary. That's assuming this is the real problem and not just a symptom of some other problem that isn't obvious from these logs.

**Comment by beqabeqa473** [comment]
I will checkout your branch now

**Comment by beqabeqa473** [comment]
Thanks @jcsteh. I can confirm, that bug is not reproduceable anymore with the way i've provided in the issue description
